### ISSUE TITLE
The useragent switcheroo module

### ISSUE THREAD

**Issue by 737simpilot** [CIDRAM/CIDRAM#193]
This is a module proposal for CIDRAM. 

Purpose: 

To catch very clever web scanners, hackers, etc.

Scope:

What this module should do is monitor if a UA changes based on a time per requests. So say you set the module to 2 seconds per 10 requests and the UA changes withen that time frame the user is either given a captcha or an out right block configurable in the module.

I'd set the time limit to be no more than 10 seconds and maybe a set amount of requests to 25 or 30. I can't see websites sending that much traffic over 30, but maybe so in which case 50 may be needed. 

For me personally, I'd set this module to catch a UA switcheroo during a season to a time limit of 2 seconds with a limit of 5 to 15 or so requests. If the UA changes withen that range and time of 2 seconds the user will get presented a captcha or is blocked. 

Evidence:

Reading my access logs going back to September to now, I have caught a handful of legit fixed line ISP IPs all withen a /24 range load two my my websites and during the first few seconds switch the UA. These were all mostly recently outdated Chrome UAs and one was HeadlessChrome. I think the common denominator between all of these requests were that they were all using HTTP 1.1 requests. Which brings me to another module idea I'll post here about. All of these so-called legit IPs from fixed line ISPs passed all of my security. Shodan and Censys has nothing on them but Grerynoise does and lists the IPs as web scanners. Two of the IPs were listed at AbuseIPDB, albeit at a very low confidence score. And I just checked now and two out of the four IPs show up at Project Honeypot.

Not sure if this can be implemented in CIDRAM or not since it would require IP and request tracking of some sort and with flat files that may be a little hard to do, but then again, maybe not. I don't know PHP, but I'm thinking the IP request could be cached along with its UA with the time of first request which would aid in this module's determination on whether the requests meet the module's settings and if not blocked or captchaed. 

Whats say you?

**Comment by Maikuolan** [comment]
Sounds like a good idea. 👍 

Not entirely sure about the logistics of it though. Due to that many browsers actively auto-update themselves while being used nowadays (e.g., Chrome), I anticipate the possibility that some users could be blocked by something like this as a result of their browser updating while in the midst of browsing a protected website. That's something we'd need to try to solve before implementing a module like this, I think. That aside though, it sounds like a cool idea.

**Comment by 737simpilot** [comment]
So with the potential of a browser auto updating during requests, I think if the settings were chosen in a way based on the time and amount of requests that would trip the module, perhaps the browser auto update avoidance can be had. 

So in my written example above, a UA would have to switch in the time frame of 2 seconds based on 5-15 requests. 5 Requests might be pretty tight and could catch a browser update. So if in the module the web Dev were to set it at 2 seconds for every 15 requests that might be lax enough to allow the possibility of a UA update all the while blocking a potential nasty. 

I guess in this sense it's kinda like a double edged sword where one setting may do the negative, in this sense the browser updating during a season, and at the same time not blocking a nasty connection. It's analogous to trying to stop a layer 7 DDoS attack. You have to set the amount of requests that will trip for an attack while at the same time not make the setting trip on a low amount of requests otherwise you'll block legit traffic. On top of that, savvy hackers can fine tune their layer 7 attack to effectively bypass your defense. I've actually heard of that happening to a user at the phpBB forum. The user had a layer 7 defense, but the attacker new to throttle it down to a point where it wouldn't trip the defense.

**Comment by 737simpilot** [comment]
Reading what I posted again, some testing on a test domain would be needed before public release I reckon. In my browser I have add-ons that can switch the UA to what ever I want, and I think I still have somewhere a very cool, but old add-on that will work in my browser to change the UA at random times. Since I use Pale Moon it uses the old tried and true XUL code for add-ons that Firefox used to use in lieu of Webextensions they now use. So right there I can thoroughly test this module's capabilities without running some server and Python script which I have no idea how to code let alone run the thing. LOL I'm pretty damn sure a lot of your bots and would-be hackers out there use Python for their shenanigans. I know this much to be true since I try to stay abreast of CVEs and what not that I find out about via various channels so I can help protect myself from it.

Edit- Just a quick search reveals this new add-on for the latest Firefox. https://addons.mozilla.org/en-US/firefox/addon/random_user_agent/ The one I had did a heck of a lot more than that, but to my dismay I can't find it now, and it isn't even in the Firefox classic archive like I thought. At the time I think I got the add-on from some website that probably didn't publish it to Mozilla's add-on database. That's just like an add-on I use now that wasn't ever in the Mozilla add-on database. It came from a website security company I think it was and after all these years still works to show me the content of the HTML5 local storage right in my browser. Google loves this storage area. Every time you go to YouTube or search Google they use your HTML5 local storage. It can be used as a form of permanent cookie if you don't clear everything out like I do upon browser exit. I even deploy the now fully free Sandboxie to take it up a notch. 

Anyway, I think if this module were created it could be well tested with timed UA changes via a browser add-on like I described. Since I saw UA switcheroo shenanigans in my logs all the while pass three layers of protection, I figured something like this may be prudent. Especially since two of the four IPs all in the same IP space show up at Greynoise, Project Honeypot and AbuseIPDB. But this guy must have blocked Shodan and Censys IPs from scanning his server. Just really curious to know how a web scanner passes the muster of Shodan and Censys. Heck, he or she is probably smart enough to block all inbound traffic on his little web scanner server. Thus the reason why only Greynoise, Project Honeypot and AbuseIPDB show him or her on radar. If these are hacked routers turned into zombie botnets, they are a pain to deal with at layer 4. It's why I submitted  the idea of a Greynoise module so that these IPs can be further cught and either out right 403ed or captchaed based on a Greynoise confidence score if their API allows it. But I don't know if the cost would merit its use. It's certainly another layer of defense though.

**Comment by 737simpilot** [comment]
I keep using the word "season." It's session. LOL

**Comment by Maikuolan** [comment]
Definitely, thorough testing prior to release would be a good idea. :-)

**Comment by Maikuolan** [comment]
So.. Just thought worth mentioning: I'm still interesting in implementing this, and it's still on the to-do list. Just been a little preoccupied with other things, plus some planning for other future changes to CIDRAM which might potentially affect this idea haven't been quite fully sketched out yet. But anyway, it'll definitely still be happening. :-)

**Comment by 737simpilot** [comment]
No problem.

**Comment by 737simpilot** [comment]
I totally forgot about this. I'll see if I can code this up with Claude after I thoroughly test a module I had built with Claude I'm using now.

**Comment by Maikuolan** [comment]
Sounds good. :-)
